tiledb/sm/filesystem/s3.cc

```diff
@@ -544,6 +544,103 @@ Status S3::flush_object(const URI& uri) {
   }
 }
 
+Status S3::finalize_and_flush_object(const URI& uri) {
+  throw_if_not_ok(init_client());
+  if (!use_multipart_upload_) {
+    throw StatusException(
```

Comment:
Make an S3StatusException class and use that when throwing exceptions so we can eventually get rid of Status_S3Error. See this for example:
class DeleteAndUpdateStatusException : public StatusException {